Tests/BSONTests/ExtendedJSONConversionTests.swift

```diff
@@ -4,7 +4,75 @@ import Nimble
 import NIO
 import XCTest
 
+extension CodingUserInfoKey {
+    static let barInfo = CodingUserInfoKey(rawValue: "bar")!
+}
+
 open class ExtendedJSONConversionTestCase: BSONTestCase {
+    struct Test: Codable, Equatable {
+        let x: Bool
+        let y: Int32
+        let z: BSONRegularExpression
+    }
+
+    struct Foo: Codable, Equatable {
+        let foo: Bool
+
+        enum CodingKeys: String, CodingKey {
+            case foo
+        }
+
+        func encode(to encoder: Encoder) throws {
+            let barInfo = encoder.userInfo[.barInfo] as! Bool
+            var container = encoder.container(keyedBy: CodingKeys.self)
+            if barInfo == true {
+                try container.encode(true, forKey: .foo)
+            } else {
+                try container.encode(self.foo, forKey: .foo)
+            }
+        }
+    }
+
+    func testExtendedJSONDecoderAndEncoder() throws {
+        let decoder = ExtendedJSONDecoder()
+        let encoder = ExtendedJSONEncoder()
+        encoder.mode = .canonical
+
+        let regexStr = "{\"$regularExpression\":{\"pattern\":\"p\",\"options\":\"i\"}}"
+        let extJSON = "{\"x\":true,\"y\":{\"$numberInt\":\"5\"},\"z\":\(regexStr)}"
+        let data = extJSON.data(using: .utf8)!
+
+        let regexObj = BSONRegularExpression(pattern: "p", options: "i")
+        let test = Test(x: true, y: 5, z: regexObj)
+
+        let decoded = try decoder.decode(Test.self, from: data)
+        let encoded: Data = try encoder.encode(test)
+
+        expect(decoded).to(equal(test))
+        let encodedStr = String(data: encoded, encoding: .utf8)
+        expect(encodedStr).to(contain("\"x\":true"))
+        expect(encodedStr).to(contain("\"y\":{\"$numberInt\":\"5\"}"))
+        expect(encodedStr).to(contain("\"z\":{\"$regularExpression\""))
+        expect(encodedStr).to(contain("\"pattern\":\"p\""))
+        expect(encodedStr).to(contain("\"options\":\"i\""))
+
+        encoder.mode = .relaxed
+        let relaxedEncoded: Data = try encoder.encode(test)
+        let relaxedEncodedStr = String(data: relaxedEncoded, encoding: .utf8)
+        expect(relaxedEncodedStr).to(contain("\"x\":true"))
+        expect(relaxedEncodedStr).to(contain("\"y\":5"))
+    }
+
+    func testExtendedJSONEncodingWithUserInfo() throws {
+        let encoder = ExtendedJSONEncoder()
+        encoder.userInfo[.barInfo] = true
+        let fooBarEncoded = try encoder.encode(Foo(foo: false))
```

Comment:
to make this test a little more straightforward, can we just remove the `foo` parameter from the `Foo` struct, and have the `encode` method just use the userInfo for encoding?

Reply:
done